### ISSUE TITLE
Issues in calling `pmem` from wasm

### ISSUE THREAD

**Issue by peterhellberg** [nesbox/TIC-80#2360]
I'm trying to write some code in Zig that depends on `pmem`, and were having some unexpected results, and then I looked in `src/api/wasm.c` and I believe there might be a bug in how the `bool` parameter to `tic_api_pmem` is assigned. _(I very rarely touch any C code so I'm a bit rusty in this regard)_

Does this mean that `writeToStorage` can never be `true`?
https://github.com/nesbox/TIC-80/blob/5061bb221fb8cd2e116efbe14295d7e3af198118/src/api/wasm.c#L579-L586

Doing something like this allows me to solve my issue, but would change the signature of `wasmtic_pmem`

```diff
diff --git a/src/api/wasm.c b/src/api/wasm.c
index 5dfd20f3..111b0655 100644
--- a/src/api/wasm.c
+++ b/src/api/wasm.c
@@ -575,16 +575,11 @@ m3ApiRawFunction(wasmtic_pmem)
     m3ApiReturnType  (uint32_t)
 
     m3ApiGetArg      (int32_t, address)
-    m3ApiGetArg      (int64_t, value)
-    bool writeToStorage;
+    m3ApiGetArg      (uint32_t, value)
+    m3ApiGetArg      (bool, writeToStorage)
 
     tic_mem* tic = (tic_mem*)getWasmCore(runtime);
 
-    // read value
-    if (value == -1) { 
-        writeToStorage = false;
-    };
-
     // TODO: this should move into tic_api_pmem, should it not?
     if (address >= TIC_PERSISTENT_SIZE) {
         m3ApiReturn(0);
@@ -987,7 +982,7 @@ M3Result linkTicAPI(IM3Module module)
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "peek4",   "i(i)",          &wasmtic_peek4)));
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "peek2",   "i(i)",          &wasmtic_peek2)));
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "peek1",   "i(i)",          &wasmtic_peek1)));
-    _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "pmem",    "i(ii)",         &wasmtic_pmem)));
+    _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "pmem",    "i(iii)",        &wasmtic_pmem)));
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "poke",    "v(iii)",        &wasmtic_poke)));
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "poke4",   "v(ii)",         &wasmtic_poke4)));
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "poke2",   "v(ii)",         &wasmtic_poke2)));
```

From the [Zig](https://ziglang.org/) :zap: side that would then be `pub extern fn pmem(index: i32, value: u32, writeToStorage: bool) u32;`

**Comment by joshgoebel** [comment]
Please look at the Lua API (and it's underlying C API) - these are the semantics we were largely trying to duplicate with the WASM API.  There are two overloads of `pmem`:

- `pmem(addr)` (read)
- `pmem(addr,value)` (write)

Since WASM has no concept of overloading (same function name, different args) there is a single pmem... `-1` is a way for the calling code to say "no 2nd argument"... ie, to call read you *at the low level* call:

`pmem(addr, -1)`

And of course you don't need to pass a value to read since the value will be returned...

---

If client libraries want to hide the detail behind a nicer abstraction (perhaps they do support overloading), then the library is welcome to do so.  That should be the libraries concern, not the APIs.  The API is purposely low-level in some things IMHO.

**Comment by peterhellberg** [comment]
Yes, the issue is that `pmem(addr, <something other than -1>)` didn't actually work :) I'll see if I can sort it out in another way.

**Comment by peterhellberg** [comment]
Another way I've found to have it work as I expect is to update the signature of `wasmtic_pmem` from `i(ii)` to `i(iI)` to have it accept the i64 in a single argument.

```diff
diff --git a/src/api/wasm.c b/src/api/wasm.c
index 5dfd20f3..6ae6b474 100644
--- a/src/api/wasm.c
+++ b/src/api/wasm.c
@@ -576,7 +576,7 @@ m3ApiRawFunction(wasmtic_pmem)
 
     m3ApiGetArg      (int32_t, address)
     m3ApiGetArg      (int64_t, value)
-    bool writeToStorage;
+    bool writeToStorage = true;
 
     tic_mem* tic = (tic_mem*)getWasmCore(runtime);
 
@@ -987,7 +987,7 @@ M3Result linkTicAPI(IM3Module module)
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "peek4",   "i(i)",          &wasmtic_peek4)));
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "peek2",   "i(i)",          &wasmtic_peek2)));
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "peek1",   "i(i)",          &wasmtic_peek1)));
-    _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "pmem",    "i(ii)",         &wasmtic_pmem)));
+    _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "pmem",    "i(iI)",         &wasmtic_pmem)));
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "poke",    "v(iii)",        &wasmtic_poke)));
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "poke4",   "v(ii)",         &wasmtic_poke4)));
     _   (SuppressLookupFailure (m3_LinkRawFunction (module, "env", "poke2",   "v(ii)",         &wasmtic_poke2)));
```

**Comment by TimotheeGreg** [comment]
This seems to be solved with #2362, it can be closed.
